### PR TITLE
Localforage implementation

### PR DESCRIPTION
**Pull request by ashish-compro**

### CODE REVIEW

File: client/app/step-builder/step-builder/step-builder.component.ts [hunk 1]
```diff
@@ -3,41 +3,49 @@ import { ActivatedRoute, Params, Router } from '@angular/router';
...
-        this.uiConfig = data;
-      });
-      let skillfilesbundle = `var skill = {}; skill.movecellcontent = {}; skill.movecellcontent.webpackBundleMap = {"moveCellContent":0}; var movecellcontentClass = (function(modules) {
+    uiConfig: UIConfig;
+    $el: any;
+    private selectedView: number;
+    taskID: string;
+    stepIndex: string;
+    builderModelSrvc;
+    constructor(el: ElementRef, private route: ActivatedRoute, private bds: BuilderDataService, private router: Router) {
```
Comment (dhirendra-singh81) [inline client/app/step-builder/step-builder/step-builder.component.ts]: previwSerive to be included

File: client/app/step-builder/step-builder/step-builder.component.ts [hunk 2]
```diff
@@ -3,41 +3,49 @@ import { ActivatedRoute, Params, Router } from '@angular/router';
...
+        this.$el = jQuery(el.nativeElement);
+        this.uiConfig = new UIConfig();
+        this.selectedView = 1;
+        this.builderModelSrvc = BuilderModelObj;
+    }
+
+    ngOnInit() {
+        localForage.config({
+            driver: localForage.INDEXEDDB,
```
Comment (dhirendra-singh81) [inline client/app/step-builder/step-builder/step-builder.component.ts]: should select driver by itself

File: client/app/step-builder/step-builder/step-builder.component.ts [hunk 3]
```diff
@@ -3,41 +3,49 @@ import { ActivatedRoute, Params, Router } from '@angular/router';
...
+            name: 'SimsBuilder',
+            version: 1.0,
+            size: 4980736, // Size of database, in bytes. WebSQL-only for now.
```
Comment (dhirendra-singh81) [inline client/app/step-builder/step-builder/step-builder.component.ts]: ?? why to limit size

File: client/app/step-builder/step-builder/step-builder.component.ts [hunk 4]
```diff
@@ -189,32 +197,69 @@ module.exports = (function () {
...
+        })
+    }
+
+    fetchSkillData() {
+        let params = {
+            id: this.taskID,
+            stepIndex: this.stepIndex
+        };
+        this.bds.getskilldata(params).subscribe((data) => {
+            this.builderModelSrvc.setModel(data["stepuistate"] || data["skillmodel"].model);
```
Comment (dhirendra-singh81) [inline client/app/step-builder/step-builder/step-builder.component.ts]: prompt to start afresh or existing one

File: client/app/step-builder/step-builder/step-builder.component.ts [hunk 5]
```diff
@@ -189,32 +197,69 @@ module.exports = (function () {
...
+            size: '6px',
+            alwaysVisible: false
+        });
+    }
+
+    checkForModelChange() {
+        let self = this;
+        let itemDataModel = this.builderModelSrvc.getModel();
+        localForage.getItem('model').then(function (value) {
+            if (JSON.stringify(value) === JSON.stringify(itemDataModel)) {
```
Comment (dhirendra-singh81) [inline client/app/step-builder/step-builder/step-builder.component.ts:244]: Is it a right way to compare ui model ?

File: server/models/taskStepUI.model.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
+const mongoose = require("mongoose");
```
Comment (dhirendra-singh81) [inline server/models/taskStepUI.model.js]: to be discarded

**Review by dhirendra-singh81 — CHANGES_REQUESTED** [review]
Update as per comments and then review with Baban

File: client/app/step-builder/shared/builder-data.service.ts [hunk 7]
```diff
@@ -15,11 +15,20 @@ export class BuilderDataService {
   constructor(private httpClient: HttpClient) { 
     this.uiconfig = new UIConfig();
   }
-  getuiconfig(params): Observable<UIConfig> {
-    return this.httpClient.get('api/skill/stepuiconfig/uiconfig/movecellcontent')
+  getskilldata(params): Observable<UIConfig> {
+    return this.httpClient.get(`api/skill/stepuiconfig/movecellcontent/${params.id}/${params.stepIndex}`)
```
Comment (vsingb3) [inline client/app/step-builder/shared/builder-data.service.ts]: params.id should be params.taskid

**Review by dhirendra-singh81 — APPROVED** [review]

**Pull request merged**